File: server/routes/groups.js
const express = require('express');
const Group = require('../models/Group');
const Message = require('../models/Message');
const User = require('../models/User');
const auth = require('../middleware/auth');
const router = express.Router();

router.post('/', auth, async (req, res) => {
  try {
    const { name, memberIds } = req.body;
    const allMembers = [...new Set([req.userId, ...(memberIds || [])])];
    const group = await Group.create({ name, admin: req.userId, members: allMembers });
    await group.populate('members', 'username handle avatar');
    res.json(group);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

router.get('/', auth, async (req, res) => {
  try {
    const groups = await Group.find({ members: req.userId })
      .populate('members', 'username handle avatar')
      .sort({ updatedAt: -1 });
    const result = await Promise.all(groups.map(async (g) => {
      const lastMsg = await Message.findOne({ groupId: g._id }).sort({ createdAt: -1 });
      return { group: g, lastMessage: lastMsg };
    }));
    res.json({ groups: result });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

router.get('/:id', auth, async (req, res) => {
  try {
    const group = await Group.findById(req.params.id).populate('members', 'username handle avatar');
    if (!group) return res.status(404).json({ error: 'Group not found' });
    res.json(group);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

router.post('/:id/members', auth, async (req, res) => {
  try {
    const { userIds } = req.body;
    const group = await Group.findById(req.params.id);
    if (!group) return res.status(404).json({ error: 'Group not found' });
    for (const uid of userIds) {
      if (!group.members.includes(uid)) group.members.push(uid);
    }
    await group.save();
    await group.populate('members', 'username handle avatar');
    res.json(group);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

router.delete('/:id/members/:memberId', auth, async (req, res) => {
  try {
    const group = await Group.findById(req.params.id);
    if (!group) return res.status(404).json({ error: 'Group not found' });
    group.members = group.members.filter(m => m.toString() !== req.params.memberId);
    await group.save();
    res.json({ success: true });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

router.get('/:id/messages', auth, async (req, res) => {
  try {
    const messages = await Message.find({ groupId: req.params.id }).sort({ createdAt: 1 });
    res.json({ messages });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

router.post('/:id/messages', auth, async (req, res) => {
  try {
    const { content } = req.body;
    const message = await Message.create({
      sender: req.userId,
      groupId: req.params.id,
      content
    });
    const io = req.app.get('io');
    if (io) io.to(`group:${req.params.id}`).emit('group_message', {
      _id: message._id,
      sender: req.userId,
      groupId: req.params.id,
      content,
      createdAt: message.createdAt
    });
    res.json(message);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: server-go/handlers/groups.go
package handlers

import (
	"context"
	"time"

	"aip-server/middleware"
	"aip-server/models"
	"aip-server/services"
	"aip-server/ws"

	"github.com/gofiber/fiber/v2"
	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

func CreateGroup(c *fiber.Ctx) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	var body struct {
		Name      string   `json:"name"`
		MemberIDs []string `json:"memberIds"`
	}
	if err := c.BodyParser(&body); err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Invalid request"})
	}

	viewerID := middleware.GetUserID(c)
	members := []bson.ObjectID{viewerID}
	seen := map[bson.ObjectID]bool{viewerID: true}

	for _, mid := range body.MemberIDs {
		oid, err := bson.ObjectIDFromHex(mid)
		if err == nil && !seen[oid] {
			members = append(members, oid)
			seen[oid] = true
		}
	}

	group := models.Group{
		ID:        bson.NewObjectID(),
		Name:      body.Name,
		Admin:     viewerID,
		Members:   members,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	if _, err := services.Groups.InsertOne(ctx, group); err != nil {
		return c.Status(500).JSON(fiber.Map{"error": err.Error()})
	}

	return c.JSON(toGroupResponse(ctx, &group))
}

func ListGroups(c *fiber.Ctx) error {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	viewerID := middleware.GetUserID(c)
	opts := options.Find().SetSort(bson.D{{Key: "updatedAt", Value: -1}})

	cursor, err := services.Groups.Find(ctx, bson.M{"members": viewerID}, opts)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": err.Error()})
	}

	var groups []models.Group
	cursor.All(ctx, &groups)

	result := make([]fiber.Map, len(groups))
	for i, g := range groups {
		var lastMsg models.Message
		services.Messages.FindOne(ctx, bson.M{"groupId": g.ID}, options.FindOne().SetSort(bson.D{{Key: "createdAt", Value: -1}})).Decode(&lastMsg)

		result[i] = fiber.Map{
			"group":       toGroupResponse(ctx, &g),
			"lastMessage": lastMsg,
		}
	}

	return c.JSON(fiber.Map{"groups": result})
}

func GetGroup(c *fiber.Ctx) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	groupID, err := bson.ObjectIDFromHex(c.Params("id"))
	if err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Invalid group ID"})
	}

	var group models.Group
	if err := services.Groups.FindOne(ctx, bson.M{"_id": groupID}).Decode(&group); err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "Group not found"})
	}

	return c.JSON(toGroupResponse(ctx, &group))
}

func AddMembers(c *fiber.Ctx) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	groupID, err := bson.ObjectIDFromHex(c.Params("id"))
	if err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Invalid group ID"})
	}

	var body struct {
		UserIDs []string `json:"userIds"`
	}
	if err := c.BodyParser(&body); err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Invalid request"})
	}

	var group models.Group
	if err := services.Groups.FindOne(ctx, bson.M{"_id": groupID}).Decode(&group); err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "Group not found"})
	}

	for _, uid := range body.UserIDs {
		oid, err := bson.ObjectIDFromHex(uid)
		if err != nil {
			continue
		}
		exists := false
		for _, m := range group.Members {
			if m == oid {
				exists = true
				break
			}
		}
		if !exists {
			group.Members = append(group.Members, oid)
		}
	}

	services.Groups.UpdateOne(ctx, bson.M{"_id": groupID}, bson.M{"$set": bson.M{"members": group.Members}})
	services.Groups.FindOne(ctx, bson.M{"_id": groupID}).Decode(&group)

	return c.JSON(toGroupResponse(ctx, &group))
}

func RemoveMember(c *fiber.Ctx) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	groupID, err := bson.ObjectIDFromHex(c.Params("id"))
	if err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Invalid group ID"})
	}

	memberID, err := bson.ObjectIDFromHex(c.Params("memberId"))
	if err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Invalid member ID"})
	}

	var group models.Group
	if err := services.Groups.FindOne(ctx, bson.M{"_id": groupID}).Decode(&group); err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "Group not found"})
	}

	newMembers := []bson.ObjectID{}
	for _, m := range group.Members {
		if m != memberID {
			newMembers = append(newMembers, m)
		}
	}
	services.Groups.UpdateOne(ctx, bson.M{"_id": groupID}, bson.M{"$set": bson.M{"members": newMembers}})

	return c.JSON(fiber.Map{"success": true})
}

func GetGroupMessages(c *fiber.Ctx) error {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	groupID, err := bson.ObjectIDFromHex(c.Params("id"))
	if err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Invalid group ID"})
	}

	opts := options.Find().SetSort(bson.D{{Key: "createdAt", Value: 1}})
	cursor, err := services.Messages.Find(ctx, bson.M{"groupId": groupID}, opts)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": err.Error()})
	}

	var messages []models.Message
	cursor.All(ctx, &messages)

	return c.JSON(fiber.Map{"messages": messages})
}

func SendGroupMessage(c *fiber.Ctx) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	groupID, err := bson.ObjectIDFromHex(c.Params("id"))
	if err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Invalid group ID"})
	}
	viewerID := middleware.GetUserID(c)

	var body struct {
		Content string `json:"content"`
	}
	if err := c.BodyParser(&body); err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Invalid request"})
	}

	message := models.Message{
		ID:        bson.NewObjectID(),
		Sender:    viewerID,
		GroupID:   &groupID,
		Content:   body.Content,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	if _, err := services.Messages.InsertOne(ctx, message); err != nil {
		return c.Status(500).JSON(fiber.Map{"error": err.Error()})
	}

	ws.H.EmitToRoom("group:"+groupID.Hex(), "group_message", fiber.Map{
		"_id":       message.ID,
		"sender":    viewerID,
		"groupId":   groupID,
		"content":   message.Content,
		"createdAt": message.CreatedAt,
	})

	return c.JSON(message)
}

func toGroupResponse(ctx context.Context, group *models.Group) fiber.Map {
	members := []fiber.Map{}
	if len(group.Members) > 0 {
		cursor, err := services.Users.Find(ctx, bson.M{"_id": bson.M{"$in": group.Members}}, options.Find().SetProjection(bson.M{"password": 0}))
		if err == nil {
			var users []models.User
			cursor.All(ctx, &users)
			for _, u := range users {
				members = append(members, toUserBrief(&u))
			}
		}
	}

	return fiber.Map{
		"id":        group.ID,
		"name":      group.Name,
		"avatar":    group.Avatar,
		"admin":     group.Admin,
		"members":   members,
		"createdAt": group.CreatedAt,
		"updatedAt": group.UpdatedAt,
	}
}
